File: backend/app/portfolio/aggregate.py
"""Portfolio aggregation — turns stored holdings into the numbers the dashboard renders.

Core idea: every holding is expanded into one or more *exposures*. A direct stock is a single
exposure; a mutual fund with look-through is expanded across its underlying buckets (so a
flexi-cap fund contributes partially to large/mid/small and a hybrid to equity+debt). All the
breakdowns (asset class, equity cap, debt, gold, sector) are then simple groupings of exposures,
which keeps the totals internally consistent.
"""

from __future__ import annotations

from dataclasses import dataclass

from sqlalchemy import select
from sqlalchemy.orm import Session, joinedload

from app.db.models import Account, Classification, Holding, Instrument
from app.domain.taxonomy import (
    ASSET_CLASS_ORDER,
    AssetClass,
    EquitySubClass,
)


def holding_value(h: Holding) -> float:
    if h.current_value is not None:
        return h.current_value
    if h.invested_value is not None:
        return h.invested_value
    return 0.0


@dataclass
class Exposure:
    holding_id: int
    instrument_id: int
    instrument_name: str
    instrument_type: str
    account: str
    source: str
    asset_class: str
    sub_class: str | None
    market_cap: str | None
    sector: str | None
    value: float
    via_lookthrough: bool
    is_estimated: bool


@dataclass
class HoldingRow:
    id: int
    instrument_id: int
    name: str
    instrument_type: str
    symbol: str | None
    isin: str | None
    scheme_code: str | None
    account: str
    source: str
    quantity: float
    avg_cost: float | None
    invested_value: float | None
    current_value: float
    last_price: float | None
    pnl: float | None
    pnl_pct: float | None
    asset_class: str
    sub_class: str | None
    market_cap: str | None
    sector: str | None
    confidence: str
    is_estimated: bool
    amc: str | None
    plan: str | None


def _load_holdings(db: Session) -> list[tuple[Holding, Instrument, Classification | None, Account]]:
    stmt = (
        select(Holding, Instrument, Classification, Account)
        .join(Instrument, Holding.instrument_id == Instrument.id)
        .join(Account, Holding.account_id == Account.id)
        .outerjoin(Classification, Classification.instrument_id == Instrument.id)
        .options(joinedload(Instrument.lookthrough))
    )
    seen: dict[int, tuple] = {}
    for row in db.execute(stmt).unique().all():
        seen[row[0].id] = row
    return list(seen.values())


def build_holding_rows(db: Session) -> list[HoldingRow]:
    rows: list[HoldingRow] = []
    for h, inst, cls, acct in _load_holdings(db):
        value = holding_value(h)
        invested = h.invested_value
        pnl = (value - invested) if (invested is not None) else None
        pnl_pct = (pnl / invested * 100) if (pnl is not None and invested) else None
        rows.append(HoldingRow(
            id=h.id, instrument_id=inst.id, name=inst.name,
            instrument_type=inst.instrument_type, symbol=inst.symbol, isin=inst.isin,
            scheme_code=inst.scheme_code, account=acct.name, source=h.source,
            quantity=h.quantity, avg_cost=h.avg_cost, invested_value=invested,
            current_value=value, last_price=h.last_price, pnl=pnl, pnl_pct=pnl_pct,
            asset_class=cls.asset_class if cls else AssetClass.UNCLASSIFIED.value,
            sub_class=cls.sub_class if cls else None,
            market_cap=cls.market_cap if cls else None,
            sector=cls.sector if cls else None,
            confidence=cls.confidence if cls else "none",
            is_estimated=cls.is_estimated if cls else True,
            amc=inst.amc, plan=inst.plan,
        ))
    return rows


def build_exposures(db: Session) -> list[Exposure]:
    exposures: list[Exposure] = []
    for h, inst, cls, acct in _load_holdings(db):
        value = holding_value(h)
        if value == 0:
            continue
        asset_class = cls.asset_class if cls else AssetClass.UNCLASSIFIED.value
        lookthrough = list(inst.lookthrough)

        if lookthrough:
            covered = 0.0
            for lt in lookthrough:
                covered += lt.weight
                exposures.append(Exposure(
                    holding_id=h.id, instrument_id=inst.id, instrument_name=inst.name,
                    instrument_type=inst.instrument_type, account=acct.name, source=h.source,
                    asset_class=lt.asset_class,
                    sub_class=lt.market_cap,
                    market_cap=lt.market_cap if lt.asset_class in (
                        AssetClass.EQUITY.value, AssetClass.INTERNATIONAL_EQUITY.value) else None,
                    sector=lt.sector,
                    value=value * lt.weight,
                    via_lookthrough=True,
                    is_estimated=lt.is_estimated,
                ))
            # Any uncovered remainder falls back to the headline asset class.
            if covered < 0.999:
                exposures.append(Exposure(
                    holding_id=h.id, instrument_id=inst.id, instrument_name=inst.name,
                    instrument_type=inst.instrument_type, account=acct.name, source=h.source,
                    asset_class=asset_class, sub_class=cls.sub_class if cls else None,
                    market_cap=cls.market_cap if cls else None,
                    sector=cls.sector if cls else None,
                    value=value * (1 - covered), via_lookthrough=True,
                    is_estimated=cls.is_estimated if cls else True,
                ))
        else:
            exposures.append(Exposure(
                holding_id=h.id, instrument_id=inst.id, instrument_name=inst.name,
                instrument_type=inst.instrument_type, account=acct.name, source=h.source,
                asset_class=asset_class, sub_class=cls.sub_class if cls else None,
                market_cap=cls.market_cap if cls else None,
                sector=cls.sector if cls else None,
                value=value, via_lookthrough=False,
                is_estimated=cls.is_estimated if cls else True,
            ))
    return exposures


# ---- grouping helpers -------------------------------------------------------------

def _group(items: list[tuple[str, float]]) -> list[dict]:
    agg: dict[str, float] = {}
    for label, value in items:
        agg[label] = agg.get(label, 0.0) + value
    total = sum(agg.values()) or 1.0
    out = [
        {"label": k, "value": round(v, 2), "pct": round(v / total * 100, 2)}
        for k, v in agg.items()
    ]
    out.sort(key=lambda d: d["value"], reverse=True)
    return out


def _ordered_group(items: list[tuple[str, float]], order: list) -> list[dict]:
    grouped = {d["label"]: d for d in _group(items)}
    out = []
    for o in order:
        label = o.value if hasattr(o, "value") else o
        if label in grouped:
            out.append(grouped.pop(label))
    out.extend(grouped.values())
    return out


def summary(db: Session) -> dict:
    rows = build_holding_rows(db)
    exposures = build_exposures(db)

    net_worth = round(sum(r.current_value for r in rows), 2)
    invested = round(sum(r.invested_value for r in rows if r.invested_value is not None), 2)
    known_value = sum(r.current_value for r in rows if r.invested_value is not None)
    pnl = round(known_value - invested, 2)
    pnl_pct = round(pnl / invested * 100, 2) if invested else 0.0
    estimated_value = round(sum(e.value for e in exposures if e.is_estimated), 2)

    by_asset = _ordered_group(
        [(e.asset_class, e.value) for e in exposures], ASSET_CLASS_ORDER
    )

    equity_exp = [e for e in exposures if e.asset_class in (
        AssetClass.EQUITY.value, AssetClass.INTERNATIONAL_EQUITY.value)]
    equity_cap = _ordered_group(
        [(_cap_label(e), e.value) for e in equity_exp],
        [s.value for s in EquitySubClass],
    )

    debt_exp = [e for e in exposures if e.asset_class == AssetClass.DEBT.value]
    debt_split = _group([(e.sub_class or "Unclassified", e.value) for e in debt_exp])

    gold_exp = [e for e in exposures if e.asset_class == AssetClass.GOLD.value]
    gold_split = _group([(e.sub_class or "Unclassified", e.value) for e in gold_exp])

    sector_exp = [e for e in exposures if e.sector and e.asset_class in (
        AssetClass.EQUITY.value, AssetClass.INTERNATIONAL_EQUITY.value)]
    sectors = _group([(e.sector, e.value) for e in sector_exp])

    by_source = _group([(r.source, r.current_value) for r in rows])
    by_account = _group([(r.account, r.current_value) for r in rows])

    return {
        "net_worth": net_worth,
        "invested": invested,
        "pnl": pnl,
        "pnl_pct": pnl_pct,
        "estimated_value": estimated_value,
        "estimated_pct": round(estimated_value / net_worth * 100, 2) if net_worth else 0.0,
        "holdings_count": len(rows),
        "asset_allocation": by_asset,
        "equity_cap_split": equity_cap,
        "debt_split": debt_split,
        "gold_split": gold_split,
        "sector_exposure": sectors,
        "by_source": by_source,
        "by_account": by_account,
    }


def _cap_label(e: Exposure) -> str:
    if e.asset_class == AssetClass.INTERNATIONAL_EQUITY.value:
        return EquitySubClass.INTERNATIONAL.value
    return e.market_cap or e.sub_class or EquitySubClass.UNCLASSIFIED.value

File: frontend/src/engine/portfolio/aggregate.ts
// Portfolio aggregation — TS mirror of backend/app/portfolio/aggregate.py.
// Each holding is expanded into exposures (funds expand across look-through), then every
// breakdown is a simple grouping of exposures, keeping all the totals internally consistent.

import { Store } from "../store/store";
import { Holding } from "../store/model";
import { ASSET_CLASS_ORDER, AssetClass, EquitySubClass } from "../taxonomy";

export function holdingValue(h: Holding): number {
  if (h.current_value != null) return h.current_value;
  if (h.invested_value != null) return h.invested_value;
  return 0;
}

export interface Exposure {
  holding_id: number;
  instrument_id: number;
  instrument_name: string;
  instrument_type: string;
  account: string;
  source: string;
  asset_class: string;
  sub_class: string | null;
  market_cap: string | null;
  sector: string | null;
  value: number;
  via_lookthrough: boolean;
  is_estimated: boolean;
}

export interface HoldingRow {
  id: number;
  instrument_id: number;
  name: string;
  instrument_type: string;
  symbol: string | null;
  isin: string | null;
  scheme_code: string | null;
  account: string;
  source: string;
  quantity: number;
  avg_cost: number | null;
  invested_value: number | null;
  current_value: number;
  last_price: number | null;
  pnl: number | null;
  pnl_pct: number | null;
  asset_class: string;
  sub_class: string | null;
  market_cap: string | null;
  sector: string | null;
  confidence: string;
  is_estimated: boolean;
  amc: string | null;
  plan: string | null;
}

const round = (n: number, d = 2) => Math.round(n * 10 ** d) / 10 ** d;
const EQUITY_CLASSES = [AssetClass.EQUITY, AssetClass.INTERNATIONAL_EQUITY];

export function buildHoldingRows(store: Store): HoldingRow[] {
  const rows: HoldingRow[] = [];
  for (const h of store.holdings) {
    const inst = store.instrument(h.instrument_id)!;
    const cls = store.classificationFor(inst.id);
    const acct = store.account(h.account_id)!;
    const value = holdingValue(h);
    const invested = h.invested_value ?? null;
    const pnl = invested != null ? value - invested : null;
    const pnl_pct = pnl != null && invested ? (pnl / invested) * 100 : null;
    rows.push({
      id: h.id, instrument_id: inst.id, name: inst.name, instrument_type: inst.instrument_type,
      symbol: inst.symbol ?? null, isin: inst.isin ?? null, scheme_code: inst.scheme_code ?? null,
      account: acct.name, source: h.source, quantity: h.quantity, avg_cost: h.avg_cost ?? null,
      invested_value: invested, current_value: value, last_price: h.last_price ?? null,
      pnl, pnl_pct,
      asset_class: cls?.asset_class ?? AssetClass.UNCLASSIFIED,
      sub_class: cls?.sub_class ?? null, market_cap: cls?.market_cap ?? null, sector: cls?.sector ?? null,
      confidence: cls?.confidence ?? "none", is_estimated: cls?.is_estimated ?? true,
      amc: inst.amc ?? null, plan: inst.plan ?? null,
    });
  }
  return rows;
}

export function buildExposures(store: Store): Exposure[] {
  const exposures: Exposure[] = [];
  for (const h of store.holdings) {
    const inst = store.instrument(h.instrument_id)!;
    const cls = store.classificationFor(inst.id);
    const acct = store.account(h.account_id)!;
    const value = holdingValue(h);
    if (value === 0) continue;
    const assetClass = cls?.asset_class ?? AssetClass.UNCLASSIFIED;
    const lookthrough = store.lookthroughFor(inst.id);
    const common = {
      holding_id: h.id, instrument_id: inst.id, instrument_name: inst.name,
      instrument_type: inst.instrument_type, account: acct.name, source: h.source,
    };
    if (lookthrough.length) {
      let covered = 0;
      for (const lt of lookthrough) {
        covered += lt.weight;
        exposures.push({
          ...common,
          asset_class: lt.asset_class, sub_class: lt.market_cap ?? null,
          market_cap: EQUITY_CLASSES.includes(lt.asset_class as any) ? lt.market_cap ?? null : null,
          sector: lt.sector ?? null, value: value * lt.weight, via_lookthrough: true, is_estimated: lt.is_estimated,
        });
      }
      if (covered < 0.999) {
        exposures.push({
          ...common,
          asset_class: assetClass, sub_class: cls?.sub_class ?? null, market_cap: cls?.market_cap ?? null,
          sector: cls?.sector ?? null, value: value * (1 - covered), via_lookthrough: true,
          is_estimated: cls?.is_estimated ?? true,
        });
      }
    } else {
      exposures.push({
        ...common,
        asset_class: assetClass, sub_class: cls?.sub_class ?? null, market_cap: cls?.market_cap ?? null,
        sector: cls?.sector ?? null, value, via_lookthrough: false, is_estimated: cls?.is_estimated ?? true,
      });
    }
  }
  return exposures;
}

export interface Slice { label: string; value: number; pct: number; }

export function group(items: [string, number][]): Slice[] {
  const agg: Record<string, number> = {};
  for (const [label, value] of items) agg[label] = (agg[label] ?? 0) + value;
  const total = Object.values(agg).reduce((a, b) => a + b, 0) || 1;
  const out = Object.entries(agg).map(([label, v]) => ({ label, value: round(v), pct: round((v / total) * 100) }));
  out.sort((a, b) => b.value - a.value);
  return out;
}

function orderedGroup(items: [string, number][], order: string[]): Slice[] {
  const grouped = new Map(group(items).map((d) => [d.label, d]));
  const out: Slice[] = [];
  for (const label of order) {
    if (grouped.has(label)) {
      out.push(grouped.get(label)!);
      grouped.delete(label);
    }
  }
  for (const d of grouped.values()) out.push(d);
  return out;
}

function capLabel(e: Exposure): string {
  if (e.asset_class === AssetClass.INTERNATIONAL_EQUITY) return EquitySubClass.INTERNATIONAL;
  return e.market_cap || e.sub_class || EquitySubClass.UNCLASSIFIED;
}
export { capLabel };

export function summary(store: Store): any {
  const rows = buildHoldingRows(store);
  const exposures = buildExposures(store);

  const net_worth = round(rows.reduce((a, r) => a + r.current_value, 0));
  const invested = round(rows.reduce((a, r) => a + (r.invested_value ?? 0), 0));
  const knownValue = rows.filter((r) => r.invested_value != null).reduce((a, r) => a + r.current_value, 0);
  const pnl = round(knownValue - invested);
  const pnl_pct = invested ? round((pnl / invested) * 100) : 0;
  const estimated_value = round(exposures.filter((e) => e.is_estimated).reduce((a, e) => a + e.value, 0));

  const byAsset = orderedGroup(exposures.map((e) => [e.asset_class, e.value] as [string, number]), ASSET_CLASS_ORDER);

  const equityExp = exposures.filter((e) => EQUITY_CLASSES.includes(e.asset_class as any));
  const equityCap = orderedGroup(equityExp.map((e) => [capLabel(e), e.value] as [string, number]), Object.values(EquitySubClass));

  const debtExp = exposures.filter((e) => e.asset_class === AssetClass.DEBT);
  const debtSplit = group(debtExp.map((e) => [e.sub_class || "Unclassified", e.value] as [string, number]));

  const goldExp = exposures.filter((e) => e.asset_class === AssetClass.GOLD);
  const goldSplit = group(goldExp.map((e) => [e.sub_class || "Unclassified", e.value] as [string, number]));

  const sectorExp = exposures.filter((e) => e.sector && EQUITY_CLASSES.includes(e.asset_class as any));
  const sectors = group(sectorExp.map((e) => [e.sector!, e.value] as [string, number]));

  const bySource = group(rows.map((r) => [r.source, r.current_value] as [string, number]));
  const byAccount = group(rows.map((r) => [r.account, r.current_value] as [string, number]));

  return {
    net_worth, invested, pnl, pnl_pct,
    estimated_value, estimated_pct: net_worth ? round((estimated_value / net_worth) * 100) : 0,
    holdings_count: rows.length,
    asset_allocation: byAsset, equity_cap_split: equityCap, debt_split: debtSplit, gold_split: goldSplit,
    sector_exposure: sectors, by_source: bySource, by_account: byAccount,
  };
}
